File: 2022/4/4.py
with open('input.txt') as f:
    data = f.readlines()

def get_sequence(elf):
    sections = list(map(int, elf.split('-')))
    return list(range(sections[0], sections[1] + 1))

def is_subsequence(seqA, seqB):
    if len(seqA) < len(seqB):
        seqA, seqB = seqB, seqA
    seqA, seqB = set(seqA), set(seqB)
    if seqA.intersection(seqB) == seqB:
        return True
    else:
        return False

def count_full_overlaps(elf_assignments):
    count = 0
    for pair in elf_assignments:
        elf1, elf2 = pair.split(',')[0], pair.split(',')[1]
        seq1, seq2, = get_sequence(elf1), get_sequence(elf2)
        if is_subsequence(seq1, seq2):
            count += 1
    return count

print(count_full_overlaps(data))

#Problem part 2

def count_inclusions(elf_assignments):
    count = 0
    for pair in elf_assignments:
        elf1, elf2 = pair.split(',')[0], pair.split(',')[1]
        seq1, seq2, = get_sequence(elf1), get_sequence(elf2)
        for area in seq1:
            if area in seq2:
                count += 1
                break
    return count

print(count_inclusions(data))

File: 2022/4/4.js
const fs = require('fs');
const input = fs.readFileSync('input.txt', 'utf-8').split('\n');

function getSequence(elf) {
    const sections = elf.split('-').map(Number);
    return Array.from({ length: sections[1] - sections[0] + 1 }, (_, i) => i + sections[0]);
}

function isSubsequence(seqA, seqB) {
    if (seqA.length < seqB.length) {
        [seqA, seqB] = [seqB, seqA];
    }
    seqA = new Set(seqA);
    seqB = new Set(seqB);
    return seqB.size === new Set([...seqA].filter(x => seqB.has(x))).size;
}

function countFullOverlaps(elfAssignments) {
    let count = 0;
    for (const pair of elfAssignments) {
        const [elf1, elf2] = pair.split(',');
        const [seq1, seq2] = [getSequence(elf1), getSequence(elf2)];
        if (isSubsequence(seq1, seq2)) {
            count++;
        }
    }
    return count;
}

console.log(countFullOverlaps(input));

// Problem part 2

function countInclusions(elfAssignments) {
    let count = 0;
    for (const pair of elfAssignments) {
        const [elf1, elf2] = pair.split(',');
        const [seq1, seq2] = [getSequence(elf1), getSequence(elf2)];
        for (const area of seq1) {
            if (seq2.includes(area)) {
                count++;
                break;
            }
        }
    }
    return count;
}

console.log(countInclusions(input));
